README.md
![jarvy](https://dl.dropboxusercontent.com/u/16169065/hal9000.jpg)
*Dave Bowman (inside EVA pod) trying to convince HAL to open the pod bay doors in “2001: A Space Odyssey” (1968).* 

	Dave: Hello, HAL. Do you read me, HAL? 
	HAL: Affirmative, Dave. I read you. 
	Dave: Open the pod bay doors, HAL. 
	HAL: I'm sorry, Dave. I'm afraid I can't do that. 

Although there has been several attempts in making the machines intelligent, the early prototypes were still dumb. Until 2015. Then came Jarvy.

### Jarvy : Python Intelligent Assistant for Humans

[![Travis](https://travis-ci.org/jarvy/jarvy.svg?branch=master)](https://github.com/jarvy/jarvy) [![PyPIv](https://img.shields.io/pypi/v/jarvy.svg)](https://pypi.python.org/pypi/jarvy) [![PyPImd](https://img.shields.io/pypi/dm/jarvy.svg)](https://pypi.python.org/pypi/jarvy)

Jarvy, aims to help humans by trying to understand them and figuring out best ways to respond to them. Jarvy is named after Tony Stark's 
artificially intelligent assistant in the Iron Man series. However, the inspiration is not restricted to the Iron Man series, but extends
 to a huge line of books, movies and projects which led to this dream. A dream of living with artificial intelligence.

By design Jarvy does not aim harm, but this is software. Who knows, what could go wrong. I will make it straight. I accept no liability, if one day Jarvy or alike overthrow humans and rule the world.

#### Why?

The goal is to have **a lot of fun**, and see where this goes. I want Jarvy to be **fully** customizable and extensible.

#### How?

    > import jarvy
    > j = jarvy.start()

#### Usage

    > Jarvy
    > Yes sir
    > who is the director of ex machina?
    > Ex Machina is a 2015 British science fiction thriller film written 
      and directed by Alex Garland, marking his directorial debut.
    > who wrote iliad?
    > Homer is best known as the author of the Iliad and the Odyssey.
    > Good bye jarvy
	> Good night sir

#### Installation

    $ pip install jarvy

#### Contribution

Hey, did you know that you can make a huge difference here? We can **together** improve Jarvy. Besides, it's more fun together! If you have **any** suggestions, please let me know. Or better, fork the repository, play with Jarvy and contribute by just sending a pull request. 

I **accept** pull requests.

#### Roadmap

Jarvy is written in Python. There is also a [web interface](https://github.com/jarvy/face) in the making. I want Jarvy to be simple 
and beautiful. Because you know, [simple is better than complex and beautiful is better than ugly](https://www.python.org/dev/peps/pep-0020/). 

#### Ideas

- Add Wolfram alpha as endpoint
- Add scoring, maybe give weights for sources?
- Add ranking to select the best response.
- Unit tests would be nice.
- Jarvy should be able learn.
- Continuous Integration would be nice.

HISTORY.rst
.. :changelog:

Release History
---------------

The latest version might be updated now and then. To keep things simpler, the latest version will override the previous one and be
incremented over unless it is a milestone.

1.3.0 (2015-07-29) [Latest Version]
++++++++++++++++++

* Handled no internet access case
* Fixed a few bugs
* Removed unnecessary package dependency
* Did some refactoring

1.2.5 (2015-07-23)
++++++++++++++++++

* Did some refactoring
* Message comparison is now in lower case
* Removed console stream for logging
* Whitespace tokenization is now employed in word matching

1.2.0 (2015-07-21)
++++++++++++++++++

* Simplified import
* Fixed bug on that prevents termination

1.1.0 (2015-07-21)
++++++++++++++++++

* Made minor changes for packaging and distribution

1.0.0 (2015-07-21)
++++++++++++++++++

* Jarvy is born!

tests.py
import unittest


class SimpleTest(unittest.TestCase):
    """Jarvy test cases."""

    def test_starting_out(self):
        self.assertEqual(1, 1)

    def setUp(self):
        import jarvy

    def tearDown(self):
        pass


def main():
    unittest.main()

if __name__ == '__main__':
    main()


jarvis/jarvis.py
import re

__author__ = 'semih'

import random
from random import randint
import time
import logging
import os

from lib.google import search
import urllib2
from bs4 import BeautifulSoup

from actions import *
from settings import *


class Jarvis:
    """ Core jarvis class
    """

    __LOG_LOCATION__ = "log/jarvis.log"

    def __init__(self):

        self.__setupLogger__()
        self.actions = Actions()
        self.settings = Settings()
        self.status = True

        self.wakeup()

    def __setupLogger__(self):

        self.logger = logging.getLogger('Jarvis')
        self.logger.setLevel(logging.INFO)

        # Log formatting
        formatter = logging.Formatter('%(asctime)s\t%(name)s\t%(levelname)s\t%(message)s')

        # Console handler
        ch = logging.StreamHandler()
        ch.setFormatter(formatter)
        self.logger.addHandler(ch)

        # Logfile handler
        directory = os.path.dirname(self.__LOG_LOCATION__)
        if not os.path.exists(directory):
            os.makedirs(directory)

        fh = logging.FileHandler(self.__LOG_LOCATION__)
        fh.setFormatter(formatter)
        self.logger.addHandler(fh)

    def wakeup(self):

        welcome = self.hi()
        print '\n\n' + welcome + "."
        self.logger.info("Woke up.")

        while self.status:
            try:
                message = str(raw_input("What can I help you with?\n"))
                answer = self.answer(message)
                print answer
            except KeyboardInterrupt:
                self.sleep()

    def sleep(self):

        farewell = self.farewell()
        print farewell
        self.status = False

        self.logger.info("Slept.")

    def hi(self):

        chance = random.uniform(0, 100)

        if chance <= 20:
            greeting = 'Hello'
        elif chance <= 40:
            greeting = 'Hi'
        else:
            hour = int(time.strftime('%H', time.localtime()))
            if 0 <= hour < 5:
                greeting = 'Good night'
            elif 5 <= hour < 12:
                greeting = 'Good morning'
            elif 12 <= hour < 17:
                greeting = 'Good afternoon'
            else:
                greeting = 'Good evening'

        return greeting

    def farewell(self):

        hour = int(time.strftime('%H', time.localtime()))
        if 0 <= hour < 5:
            farewell = 'Good night'
        elif 5 <= hour < 12:
            farewell = 'Have a good day'
        elif 12 <= hour < 17:
            farewell = 'See you again'
        else:
            farewell = 'Good night'

        return farewell

    def answer(self, message):

        action = self.understand(message)
        options = self.think(action, message)
        answer = self.explain(options, message)

        return answer

    def understand(self, message):

        if any(m in message for m in self.settings.personal_message_for_jarvis):    # it is something about the jarvis
            action = self.actions.about_jarvis
        elif message == self.settings.jarvis_name:
            action = self.actions.direct_address
        elif any(m in message for m in self.settings.personal_message_for_master):   # it is something about the master
            action = self.actions.about_master
        elif any(m in message for m in self.settings.rudimentary_question_tags):     # we can rewrite the message as a search query
            action = self.actions.search_google
        else:
            action = self.actions.say_sorry
        return action

    def think(self, action, message):

        if action == self.actions.about_jarvis:
            options = self.get_options_for_personal_questions('jarvis', message)
        elif action == self.actions.direct_address:
            options = ['Yes ' + self.settings.master_formal_address]
        elif action == self.actions.about_master:
            options = self.get_options_for_personal_questions('master', message)
        elif action == self.actions.search_google:  # get results from google
            options = self.make_search(message, 'google')
        elif action == self.actions.search_wolfram:  # get results from wolfram
            options = self.make_search(message, 'wolfram')
        elif action == self.actions.search_wikipedia:  # get results from wikipedia
            options = self.make_search(message, 'wikipedia')
        else:
            rand = randint(0, len(self.settings.sorry_messages) - 1)
            options = [self.settings.sorry_messages[rand]]
        return options

    def make_search(self, query, source):

        query = self.rewrite_question(source, query)

        search_results = []
        trusted_results = []
        urls = search(query, stop=10)  # make a search

        for i, url in enumerate(urls):
            search_results.append(url)
            if url.find('wikipedia') > 0 and i < 3:  # trusted source, more likely to include answer, skip if not in the top 3
                trusted_results.append(url)
        try:

            if len(trusted_results) > 0:
                html = urllib2.urlopen(trusted_results[0]).read()
                soup = BeautifulSoup(html, 'html.parser')
                p = soup.find('div', id="bodyContent").p
                soup_p = BeautifulSoup(str(p), 'html.parser')
                text = soup_p.get_text()
            else:
                # print search_results[0]
                html = urllib2.urlopen(search_results[0]).read()
                soup = BeautifulSoup(html, 'html.parser')
                p = soup.find_all('p')
                soup_p = BeautifulSoup(str(p), 'html.parser')
                text = soup_p.get_text()
                pat = re.compile(r'([A-Z][^\.!?]*[\.!?])', re.M)  # pattern: Uppercase, and anything that is not in (.!?), and one of them
                matches = pat.findall(text)
                if len(matches) > 3:
                    text = ''.join(matches[0:3])
                else:
                    text = ''.join(matches[0:])
        except:
            text = ''

        options = [text]
        return options


    def explain(self, options, message):

        if len(options) >= 1:
            answer = options[0]
        else:
            rand = randint(0, len(self.settings.sorry_messages) - 1)
            answer = self.settings.sorry_messages[rand]
        return answer

    def get_options_for_personal_questions(self, person, message):

        if person == 'jarvis':
            options = ['Let\'s not talk about me.']
        elif person == 'master':
            options = ['You know, I can\'t answer that']
        else:
            options = ['']

        return options

    def rewrite_question(self, source, message):

        query = message  # by default, set it to message

        if source == 'google':
            for r in self.settings.rudimentary_question_tags:
                query = query.replace(r, '')

            query = query.replace('is', '')
            query = query.replace('are', '')
            query = query.strip()

        return query


def main(self):

    jarvis = Jarvis()
    return 0

if __name__ == '__main__':
    main(None)


jarvy/settings.py


class Settings:

    def __init__(self):
        pass

    # default settings

    master_title = 'Mr.'
    master_name = 'Semih'
    master_surname = 'Yagcioglu'
    master_gender = 'female'
    master_formal_address = 'sir'
    master_email_username = None
    master_email_password = None
    jarvy_name = 'jarvy'
    jarvy_gender = 'female'

    # messages
    sorry_messages = ['I am sorry, I could not understand that', 'Oh oh',
                      'I am afraid, I can\'t talk about that',
                      'I am thinking but this beats me...', 'Allright allright, I give up', 'Who knows',
                      'Don\'t get me wrong but I don\'t know all the answers',
                      'Hey, why don\'t you ask me another question while I think more about that?',
                      'Believe me I am trying but, I simply could not answer that',
                      'I think I do not understand you',
                      'Hmmm, I am doing my best,just be patient, OK?']
    farewell_messages = ['good bye', 'bye', 'see you', 'farewell']
    positive_answers = ['Yes', 'Of course', 'Definitely']
    negative_answers = ['No', 'Nope', 'I don\'t think so', 'Sorry']
    personal_message_for_jarvy = ['you', 'yourself']
    personal_message_for_master = ['me', 'myself', 'I']
    rudimentary_question_tags = ['who', 'why', 'when', 'where', 'what', 'which', 'how']

    # algorithm parameters
    number_of_search_results = 10
    trusted_source_treshold = 3
    number_of_minimum_sentences = 3  # should reconsider this




jarvy/__init__.py
import re
import random
from random import randint
import time
import logging
import os
import urllib2
from bs4 import BeautifulSoup

from packages.google import search

from actions import *
from settings import *

"""
Jarvy : Python Intelligent Assistant for Humans
=================

Jarvy, aims to help humans by trying to understand them and figuring out best ways to respond to them.

"""

__title__ = 'jarvy'
__version__ = '1.3.0'
__build__ = 0x010300  # in the format of 00-00-00
__author__ = 'Semih Yagcioglu'
__license__ = 'MIT'
__copyright__ = 'Copyright 2015 Semih Yagcioglu'


class Jarvy:
    """ Jarvy : Python Intelligent Assistant for Humans
    """

    __LOG_LOCATION__ = "log/jarvy.log"

    def __init__(self):

        self.__setupLogger__()
        self.actions = Actions()
        self.settings = Settings()
        self.status = True

        self.wakeup()

    def __setupLogger__(self):

        self.logger = logging.getLogger('Jarvy')
        self.logger.setLevel(logging.INFO)

        # Log formatting
        formatter = logging.Formatter('%(asctime)s\t%(name)s\t%(levelname)s\t%(message)s')

        # Console handler
        # ch = logging.StreamHandler()
        # ch.setFormatter(formatter)
        # self.logger.addHandler(ch)

        # Logfile handler
        directory = os.path.dirname(self.__LOG_LOCATION__)
        if not os.path.exists(directory):
            os.makedirs(directory)

        fh = logging.FileHandler(self.__LOG_LOCATION__)
        fh.setFormatter(formatter)
        self.logger.addHandler(fh)

    # starts jarvy
    def wakeup(self):

        welcome = self.hi()
        print '\n\n' + welcome + "."
        self.logger.info("Woke up.")

        while self.status:
            try:
                message = str(raw_input("What can I help you with?\n\n")).lower()

                if any(m in message for m in self.settings.farewell_messages) and any(m in message for m in self.settings.jarvy_name):
                    self.sleep()
                else:
                    answer = self.answer(message)
                    print answer
            except KeyboardInterrupt:
                self.sleep()

    # terminates jarvy
    def sleep(self):

        farewell = self.farewell()
        print farewell
        self.status = False

        self.logger.info("Slept.")

    # produce salutation message
    def hi(self):

        chance = random.uniform(0, 100)

        if chance <= 20:
            greeting = 'Hello ' + self.settings.master_name
        elif chance <= 30:
            greeting = 'Hi ' + self.settings.master_name
        elif chance <= 40:
            greeting = 'Hey'
        else:
            hour = int(time.strftime('%H', time.localtime()))
            if 0 <= hour < 5:
                greeting = 'Good night ' + self.settings.master_formal_address
            elif 5 <= hour < 12:
                greeting = 'Good morning ' + self.settings.master_formal_address
            elif 12 <= hour < 17:
                greeting = 'Good afternoon ' + self.settings.master_formal_address
            else:
                greeting = 'Good evening ' + self.settings.master_formal_address

        return greeting

    # produce farewell message
    def farewell(self):

        chance = random.uniform(0, 100)

        if chance <= 20:
            farewell = 'Good bye ' + self.settings.master_name
        elif chance <= 40:
            farewell = 'Farewell my friend'
        else:
            hour = int(time.strftime('%H', time.localtime()))
            if 0 <= hour < 5:
                farewell = 'Good night ' + self.settings.master_formal_address
            elif 5 <= hour < 12:
                farewell = 'Have a good day ' + self.settings.master_formal_address
            elif 12 <= hour < 17:
                farewell = 'Good afternoon ' + self.settings.master_formal_address
            else:
                farewell = 'Good evening ' + self.settings.master_formal_address

        return farewell

    # respond to the query message
    def answer(self, message):

        action = self.understand(message)
        responses = self.think(action, message)
        answer = self.explain(responses)

        return answer

    def understand(self, message):

        if any(m in message.split() for m in self.settings.personal_message_for_jarvy):    # it is something about the jarvy
            action = self.actions.about_jarvy
        elif message == self.settings.jarvy_name.lower():  # direct address to jarvy
            action = self.actions.direct_address
        elif any(m.lower() in message.split() for m in self.settings.personal_message_for_master):   # it is something about the master
            action = self.actions.about_master
        elif any(m.lower() in message.split() for m in self.settings.rudimentary_question_tags):     # rewrite message as a search query
            action = self.actions.search_google
        else:
            action = self.actions.say_sorry
        return action

    def think(self, action, message):

        if action == self.actions.about_jarvy:
            responses = self.respond('jarvy')
        elif action == self.actions.direct_address:
            responses = self.respond('direct_address')
        elif action == self.actions.about_master:
            responses = self.respond('master')
        elif action == self.actions.search_google:  # get results from google
            responses = self.make_search(message, 'google')
        elif action == self.actions.search_wolfram:  # get results from wolfram
            responses = self.make_search(message, 'wolfram')
        elif action == self.actions.search_wikipedia:  # get results from wikipedia
            responses = self.make_search(message, 'wikipedia')
        else:
            responses = []

        if not responses:
            rand = randint(0, len(self.settings.sorry_messages) - 1)
            responses = [self.settings.sorry_messages[rand]]

        return responses

    def make_search(self, query, source):

        query = self.rewrite_question(source, query)

        search_results = []
        trusted_results = []

        try:
            urls = search(query, stop=self.settings.number_of_search_results)  # make a search

            for i, url in enumerate(urls):
                search_results.append(url)
                if url.find('wikipedia') > 0 and i < self.settings.trusted_source_treshold:  # trusted source, skip if not in there
                    trusted_results.append(url)
            try:
                if len(trusted_results) > 0:
                    html = urllib2.urlopen(trusted_results[0]).read()
                    soup = BeautifulSoup(html, 'html.parser')
                    p = soup.find('div', id="bodyContent").p
                    soup_p = BeautifulSoup(str(p), 'html.parser')
                    text = soup_p.get_text()
                else:
                    html = urllib2.urlopen(search_results[0]).read()
                    soup = BeautifulSoup(html, 'html.parser')
                    p = soup.find_all('p')
                    soup_p = BeautifulSoup(str(p), 'html.parser')
                    text = soup_p.get_text()
                    pat = re.compile(r'([A-Z][^\.!?]*[\.!?])', re.M)  # pattern to detect sentences
                    matches = pat.findall(text)

                    if len(matches) > self.settings.number_of_minimum_sentences:
                        text = ''.join(matches[0:self.settings.number_of_minimum_sentences])
                    else:
                        text = ''.join(matches[0:])  # TODO: what if the answer is = ''

                responses = [text]
            except:
                responses = []  # there is no proper response, but need to reconsider this
        except:
            responses = []  # no internet access

        return responses

    def explain(self, responses):

        if len(responses) >= 1:  # if there is only one response
            answer = responses[0]
        else:
            pass  # TODO: scoring might be done here!

        return answer

    def respond(self, about):

        if about == 'jarvy':
            responses = ['I am your friend.']
        elif about == 'direct_address':
            responses = ['Yes ' + self.settings.master_formal_address]
        elif about == 'master':
            responses = ['I\'m afraid, I can\'t answer that.']
        else:
            responses = ['']

        return responses

    def rewrite_question(self, source, message):

        query = message  # by default, set it to message

        if source == 'google':
            for r in self.settings.rudimentary_question_tags:
                query = query.replace(r, '')

            query = query.replace(' is ', ' ')
            query = query.replace(' are ', ' ')
            query = query.strip()

        return query


def start():

    j = Jarvy()
    return j

if __name__ == '__main__':
    start()

